native/python/zds_py.cpp

```diff
@@ -0,0 +1,210 @@
+#include "zds_py.hpp"
+#include <unistd.h>
+#include <string>
+#include <stdexcept>
+
+static void e2a_inplace(std::string &s)
+{
+    if (s.empty())
+        return;
+    s.push_back('\0');
+    __e2a_s(&s[0]);
+    s.pop_back();
+}
+
+static void a2e_inplace(std::string &s)
+{
+    if (s.empty())
+        return;
+    s.push_back('\0');
+    __a2e_s(&s[0]);
+    s.pop_back();
+}
+
+/**
+ * Dataset Functions
+ */
+
+void create_dataset(std::string dsn, DS_ATTRIBUTES attributes)
+{
+    ZDS zds = {0};
+    std::string response;
+
+    DS_ATTRIBUTES attrs_copy = attributes;
+
+    a2e_inplace(dsn);
+
+    a2e_inplace(attrs_copy.alcunit);
+    a2e_inplace(attrs_copy.dsorg);
+    a2e_inplace(attrs_copy.recfm);
+    a2e_inplace(attrs_copy.dataclass);
+    a2e_inplace(attrs_copy.unit);
+    a2e_inplace(attrs_copy.dsntype);
+    a2e_inplace(attrs_copy.mgntclass);
+    a2e_inplace(attrs_copy.dsname);
+    a2e_inplace(attrs_copy.storclass);
+    a2e_inplace(attrs_copy.vol);
+
+    int rc = zds_create_dsn(&zds, dsn, attrs_copy, response);
+
+    if (rc != 0)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+}
+
+std::vector<ZDSEntry> list_datasets(std::string dsn)
+{
+    std::vector<ZDSEntry> entries;
+    ZDS zds = {0};
+
+    a2e_inplace(dsn);
+    int rc = zds_list_data_sets(&zds, dsn, entries);
+
+    if (rc != 0 && rc != RTNCD_WARNING)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+
+    for (auto &e : entries)
+    {
+        e2a_inplace(e.name);
+        e2a_inplace(e.dsorg);
+        e2a_inplace(e.volser);
+        e2a_inplace(e.recfm);
+    }
+    return entries;
+}
+
+std::string read_dataset(std::string dsn, std::string codepage)
+{
+    ZDS zds = {0};
+
+    if (!codepage.empty())
+    {
+        zds.encoding_opts.data_type = eDataTypeText;
+        strncpy(zds.encoding_opts.codepage, codepage.c_str(), sizeof(zds.encoding_opts.codepage) - 1);
+    }
+
+    a2e_inplace(dsn);
+    std::string response;
+    int rc = zds_read_from_dsn(&zds, dsn, response);
+
+    if (rc != 0)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+
+    e2a_inplace(response);
+    return response;
+}
+
+std::string write_dataset(std::string dsn, std::string data, std::string codepage, std::string etag)
```

Comment:
- Can a user specify `binary` when reading from/writing to a data set?
- `data_type` for `zds.encoding_opts` should be set to `eDataTypeBinary` when a binary read or write operation is requested